addon/index.d.ts

```diff
@@ -0,0 +1,940 @@
+import ComputedProperty from '@ember/object/computed';
+
+export type TaskGenerator<T> = Generator<any, T, any>;
+
+export type TaskFunction<T, Args extends any[]> = (...args: Args) => TaskGenerator<T>;
+
+export type TaskFunctionArgs<T extends TaskFunction<any, any[]>> =
+  T extends (...args: infer A) => TaskGenerator<any> ? A : [];
+
+export type TaskFunctionReturnType<T extends TaskFunction<any, any[]>> =
+  T extends (...args: any[]) => TaskGenerator<infer R> ? R : unknown;
+
+export interface EncapsulatedTaskDescriptor<T, Args extends any[]> {
+  perform(...args: Args): TaskGenerator<T>;
+}
+
+export type EncapsulatedTaskDescriptorArgs<T extends EncapsulatedTaskDescriptor<any, any[]>> =
+  T extends { perform(...args: infer A): TaskGenerator<any> } ? A : [];
+
+export type EncapsulatedTaskDescriptorReturnType<T extends EncapsulatedTaskDescriptor<any, any[]>> =
+  T extends { perform(...args: any[]): TaskGenerator<infer R> } ? R : unknown;
```

Comment:
Because the feature is called and documented as `EncapsulatedTask`. I considered just calling it `EncapsulatedTask `, but that's not quite right, because this is the thing you pass in (similar to the POJO you pass to `EmberObject.extend(...)`) to create a encapsulated task (technically, an encapsulated task property, which gets turned into an encapsulated task, which can be instantiated as encapsulated task instances). It just so happens that the current implementation of the `task` function normalizes away the "encapsulated-ness" of it so there aren't `EncapsulatedTaskProperty`, `EncapsulatedTask` and `EncapsulatedTaskInstance` exports, but there very well may be in the future.

As far as `TaskDescriptor`, that also sounds like too generic of a term, as it seems feasible that there will be a `TaskDescriptor` that is unrelated to `EncapsulatedTask`.

Ultimately, I question whether some of these APIs are still needed/still a good fit in the post-Octane world, but ultimately, it just doesn't seem like my decision to make or my job to speculate on the future direction too much in this PR, as my goal is just to document what exists today.

So `EncapsulatedTaskDescriptorArgs` seems like the most conservative, and accurate, if a bit verbose, option.

Reply:
This makes sense; thanks for the explanation! I had forgotten about the feature in question. 👍